bundle/bundle.go

```diff
@@ -327,37 +329,52 @@ func (b Bundle) Equal(other Bundle) bool {
 }
 
 func (b *Bundle) insert(key []string, value interface{}) error {
-	if len(key) == 0 {
-		obj, ok := value.(map[string]interface{})
-		if !ok {
-			return fmt.Errorf("root value must be object")
-		}
-		b.Data = obj
-		return nil
-	}
-
-	obj, err := b.mkdir(key[:len(key)-1])
+	// Build an object with the full structure for the value
+	obj, err := mktree(key, value)
 	if err != nil {
 		return err
 	}
 
-	obj[key[len(key)-1]] = value
+	// Merge the new data in with the current bundle data object
+	merged, ok := merge.InterfaceMaps(b.Data, obj)
+	if !ok {
+		return fmt.Errorf("failed to insert data file from path %s", filepath.Join(key...))
+	}
+
+	b.Data = merged
+
 	return nil
 }
 
-func (b *Bundle) mkdir(key []string) (map[string]interface{}, error) {
-	obj := b.Data
-	for i := 0; i < len(key); i++ {
-		node, ok := obj[key[i]]
+func mktree(key []string, value interface{}) (map[string]interface{}, error) {
+	if len(key) == 0 {
+		// For 0 length keys the value is the full tree.
+		obj, ok := value.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("root value must be object")
+		}
+		return obj, nil
+	}
+
+	dir := key[:len(key)-1]
+
+	obj := map[string]interface{}{}
+	next := obj
+	for i := 0; i < len(dir); i++ {
+		node, ok := next[key[i]]
```

Comment:
Nit: this works but the !ok expressions on line 365 and 370 are always true and never true (respectively). If you construct the tree inside-out then you don't need these type assertions. For example: https://play.golang.org/p/b327J4c9vjw

Reply:
Thats a cool idea. I had left the original algorithm here pretty much intact aside from keeping a reference to the top level map, but will switch to this approach.